### PR TITLE
[weight checker] refactor: add precision branch; allow ULP quant err; used chunked compare

### PR DESCRIPTION
**Pull request by yueming-yuan**
## Motivation

The weight checker's `compare`/`checksum` actions require two block-quantized weights to be **bit-identical** after dequant. But two independent yet faithful quantizations of the same source weight — e.g. before vs. after a weight update that requantizes — can legitimately differ per element. Tolerating that with a hand-tuned mean-error threshold is both too loose (a mean lets per-element corruption hide in a large tensor) and arbitrary per model/precision.

This replaces it with a principled, per-element tolerance: each side of a block-quant pair is one faithful quantization of the same source weight, so each may deviate from it by up to 1 ULP of its own representation — i.e. the two may differ by at most `expect_ulp + actual_ulp` per element in dequantized space. Opt in via `allow_quant_error`; the default stays exact-equality (no behavior change).

## Modifications

- `CheckWeightsReqInput` gains `allow_quant_error: bool = False`, plumbed through scheduler / `model_runner.check_weights` into the weight checker.
- New `weight_checker_quant.py` holds the quantization-aware comparison behind a small `ReferenceWeight` seam (`Fp8BlockReference` today). Adding a precision is a new subclass.
- `select_quantization_method` dispatches on `module.quant_method` (not param names — robust to swizzled scales / per-format naming) and raises `NotImplementedError` for formats with no `ReferenceWeight`, so mxfp8 / nvfp4 / int4 fail loudly instead of being silently mis-compared.
- Comparison/dequant is chunked to bound GPU memory; NaN counts as exceeding tolerance (`~(diff <= tol)`); block size is inferred from weight/scale shapes (handles partial last blocks and non-128 block formats).

## Tests

`test/registered/unit/utils/test_weight_checker.py` (CUDA-gated):
- `TestQuantUlp` — `_quant_ulp` brute-force verified against the true spacing of all 256 bit patterns of `float8_e4m3fn` / `float8_e5m2` (exact match).
- `TestCompareQuantPair` — two quantizations within tolerance pass; bit-flip corruption and fp8 NaN exceed; chunked == unchunked; partial last block; 3D expert tensors; ue8m0-packed scales.
- `TestSelectQuantizationMethod` — unsupported quant method raises.

































































































































































































































































<!-- pr-states:start -->
---
### CI States

Latest PR Test (Base): <!-- slot:pr-test:start -->:no_entry_sign: [Run #28343424880](https://github.com/sgl-project/sglang/actions/runs/28343424880)<!-- slot:pr-test:end -->
Latest PR Test (Extra): <!-- slot:pr-test-extra:start -->:x: [Run #28343424787](https://github.com/sgl-project/sglang/actions/runs/28343424787)<!-- slot:pr-test-extra:end -->
<!-- pr-states:end -->

### CODE REVIEW

File: python/sglang/srt/utils/weight_checker.py [hunk 1]
```diff
@@ -223,58 +242,42 @@ def _random_like(t: torch.Tensor):
...
+
+def _build_entries(
     raw: Dict[str, torch.Tensor],
     skip_compare_names: Set[str],
-) -> Iterable[Tuple[str, bool, torch.Tensor]]:
-    from sglang.srt.debug_utils.dumper import get_tensor_info
-
+    quantized_set: Optional[Dict[str, Tuple[type, str]]] = None,
+) -> Iterable[Tuple[str, bool, ComparableWeight]]:
+    """Yields (name, should_compare, ComparableWeight); quantized weights consume
```
Comment (fzyzcjy) [inline python/sglang/srt/utils/weight_checker.py]: nit: use dataclass / NamedTuple everywhere to be slightly more readable

File: python/sglang/srt/utils/weight_checker.py [hunk 2]
```diff
@@ -7,11 +7,14 @@
 import torch.distributed as dist
 from pydantic import BaseModel, ConfigDict
 
-from sglang.srt.layers.quantization.fp8_utils import (
-    block_quant_dequant,
-    inverse_transform_scale_ue8m0,
-)
 from sglang.srt.managers.mm_utils import tensor_hash
+from sglang.srt.utils.weight_checker_comparator import (
+    _CHUNK_NUMEL,
```
Comment (fzyzcjy) [inline python/sglang/srt/utils/weight_checker.py]: nit: do not import private things

File: test/registered/unit/utils/test_weight_checker.py [hunk 3]
```diff
@@ -328,30 +343,258 @@ def test_raises_when_should_compare_true_and_diff(self):
...
+            ("a", True, RawComparable(t.clone())),
+            ("b", True, RawComparable(t.clone())),
+        ]
+        actual = [("a", True, RawComparable(t.clone()))]
         with self.assertRaises(ValueError):
             _check_tensors(expect_tensors=expect, actual_tensors=actual)
 
 
+# ---------------------------------------------------------------------------
+# _quant_ulp + allow_quant_error
```
Comment (fzyzcjy) [inline test/registered/unit/utils/test_weight_checker.py]: nit: test_weight_checker_comparator.py

**Review by fzyzcjy — COMMENTED** [review]
general direction LGTM, not check details though

**Review by fzyzcjy — APPROVED** [review]

**Review by hnyls2002 — APPROVED** [review]

File: python/sglang/srt/utils/weight_checker.py
Comment (fzyzcjy) [inline python/sglang/srt/utils/weight_checker.py:1]: random thought: let init hf weight be hacked into all nan to detect local attn buf issue
